## Question 1 [20 points] 
## In no more than 500 words, explain the differences between a Jupyter Notebook (.ipynb extension) and a Python file (.py extension). Focus on the advantages and disadvantages of one or the other, particularly for data analysts/scientists. You are encouraged to use examples and screenshots of the different scripts to support your arguments. 

While Python files can be read and edited directly in an IDE, such as Pycharm, Jupyter Notebooks must be accessed by running the command ‘jupyter notebook’ in the terminal, from which a url will be generated, taking the user to the Jupter Notebook in webpage. 

Jupyter notebooks are best for presenting and explaining code, as the code can be broken down into small segments (cells). The code can also be run in any order according to the segments, which can allow for demonstrations of variations in the code. In addition, the Jupyter Notebook format supports different text types (‘Markdown’, ‘Code’, ‘Heading’, ‘Raw NBConvert’)  with text and images being able to be integrated throughout the notebook. In contrast, Python files are run as the full code in the file, unless code is commented out using the # key. Notes in Python files separate from the actual code must be commented out using # or “””. When code is run, the output appears directly below the cell that has been run in Jupyter, or in the Run tool window in Python. It should be noted that for Jupyter the output appears immediately, whereas in order to get an output in Python the programmer must use print(). 

In my experience, Python files are better for researching, working on and developing code, and Jupyter notebooks are best for presenting and explaining code. In this way, Python files are best used by data and software engineers, and Jupyter files are an appropriate tool for data analysts and data scientists when presenting their program/code to lay colleagues.  


## Question 2 [10 points] 
## What is the difference between a Pandas DataFrame and a Pandas series. Show an example of how you create each of them. 

A Panda series is a column of data. A Pandas DataFrame is a collection of Pandas series. Combining more than one Pandas series results in a Pandas DataFrame. A Pandas DataFrame is presentation of the data in rows and columns. To create a Pandas DataFrame use the syntax:


In [ ]:
import pandas as pd
DataFrame = pd.read_csv(‘file location’)
DataFrame.head()

The final line (DataFrame.head()) will give you the first five and last five rows of the data. 

To create a Pandas Series:

In [ ]:
import pandas as pd

example = ['first_item', 'second_item', 'third_item']
pd.Series(example)

## Question 3 [10 points] 
## Starting from the argument that a Pandas DataFrame represents rectangular data, use the internet and other resources to describe in no more than a few sentences the difference between rectangular and non-rectangular data. 

As explained on (https://bookdown.org/hneth/ds4psy/1.5-basics:tables.html), each record in a Pandas DataFrame is a ‘one-dimensional […] linear vector’. When combined in the DataFrame, the data becomes ‘two-dimensional’, as it is now in rows and columns. This is a rectangular data structure. Non-rectangular data is not ordered in the same way. It is made up of multiple data structures, normally stored in lists. 

## Question 4 [10 points] 
## Starting from the data visualisation usage from Session 2, give examples of when figures could be: 
### 1.	Of use to the data scientist to identify patterns in the data/highlight the important parts of a data set:
Scattergrams are very good for showing whether two variables have a relationship, and showing distinct patterns. They are also useful for highlighting outliers. Similarly, the whiskers on box plots are very useful for assessing the reliability of data. Box plots also communicate clearly descriptive statistics as part of the data, by showing the mean, interquartile range, and skewness visually.

### 2.	Tell a story and create business presentations. 
Line charts and bar charts are useful for business presentations. Line charts are excellent for showing trends in data, particularly over a period of time. The axis can begin at any point, so they can be very useful in illustrating specific changes in the data. Bar charts are very simple and clear. 

## Question 5 [50 points]
## Each sub-question is worth 10 points. 


In [1]:
import pandas as pd
titanic = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')
titanic 

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

## 1. How many columns and rows does the data have? 
15 columns
891 rows


In [ ]:
titanic.shape

## 2. Get a sense of your data and find the min, max, and count/mean depending on the data type. 

In [ ]:
titanic['age'].min()
# youngest Titanic passenger (in years)

In [ ]:
titanic['age'].max()
# oldest Titanic passenger (in years)

In [ ]:
titanic['fare'].mean()

In [ ]:
titanic.describe()

## 3.	Give an overview (code and an explanation) of all missing values in the data. 

In [ ]:
len(titanic)-titanic.count()

In [ ]:
titanic.age.isna().count()

In [ ]:
titanic.age.count()

## 4.	Delete the rows where you do not have information about the age of the person. Then group the passengers in a 10 year age range (for example, you can do something like 0 – 10, 11 – 20, 21 – 30, etc). 

In [2]:
test_titanic = titanic.copy()
test_titanic.shape

(891, 15)

In [3]:
test_titanic.dropna(subset=['age'], inplace=True)

In [ ]:
test_titanic.shape

In [4]:
def age_function(age):
    if age >= 0 and age < 10:
        AgeGroup = '0-9'
    elif age >= 10 and age < 20:
        AgeGroup = '10-19'
    elif age >= 20 and age < 30:
        AgeGroup = '20-29'
    elif age >= 30 and age < 40:
        AgeGroup = '30-39'
    elif age >= 40 and age < 50:
        AgeGroup = '40-49'
    elif age >= 50 and age < 60:
        AgeGroup = '50-59'
    elif age >= 60 and age < 70:
        AgeGroup = '60-69'
    elif age >= 70 and age < 80:
        AgeGroup = '70-79'
    else:
        AgeGroup = '80+'
    return AgeGroup

test_titanic['age_range'] = titanic['age'].apply(age_function)
age_range_var = test_titanic.groupby('age_range')['age'].count()
age_range_var

age_range
0-9       62
10-19    102
20-29    220
30-39    167
40-49     89
50-59     48
60-69     19
70-79      6
80+        1
Name: age, dtype: int64

In [16]:

females = [test_titanic['sex']]=='female'
males = [test_titanic['sex']]=='male'
first_class = [test_titanic['class']]=='First'
second_class = [test_titanic['class']]=='Second'
third_class = [test_titanic['class']]=='Third'




